Framework/DataObjects/src/TimeSplitter.cpp

```diff
@@ -280,5 +293,171 @@ TimeROI TimeSplitter::getTimeROI(const int workspaceIndex) {
 
 std::size_t TimeSplitter::numRawValues() const { return m_roi_map.size(); }
 
+// ------------------------------------------------------------------------
+// SPLITTING EVENTS METHODS
+// ------------------------------------------------------------------------
+
+/**
+ * Split a list of events according to Pulse time or Pulse + TOF time
+ *
+ * Events with masked times are allocated to destination index -1.
+ * @param events : list of input events
+ * @param partials : resulting partial lists of events
+ * @param pulseTof : if True, split according to Pulse + TOF time, otherwise split by Pulse time
+ * @param tofCorrect : rescale and shift the TOF values (factor*TOF + shift)
+ * @param factor : rescale the TOF values by a dimensionless factor.
+ * @param shift : shift the TOF values after rescaling, in units of microseconds.
+ * @raises RunTimeError : the event list is of type Mantid::API::EventType::WEIGHTED_NOTIME
+ */
+void TimeSplitter::splitEventList(const EventList &events, std::map<int, EventList *> partials, bool pulseTof,
+                                  bool tofCorrect, double factor, double shift) const {
+
+  if (events.getEventType() == EventType::WEIGHTED_NOTIME)
+    throw std::runtime_error("EventList::splitByTime() called on an EventList "
+                             "that no longer has time information.");
+
+  this->initializePartials(events, partials);
+
+  if (this->empty())
+    return;
+
+  // sort the list of events and return their times, in nanoseconds
+  std::vector<int64_t> times = this->sortEventList(events, pulseTof, tofCorrect, factor, shift);
+
+  switch (events.getEventType()) {
+  case EventType::TOF:
+    this->splitEventVec(times, events.getEvents(), partials);
+    break;
+  case EventType::WEIGHTED:
+    this->splitEventVec(times, events.getWeightedEvents(), partials);
+    break;
+  default:
+    throw std::runtime_error("Unhandled event type");
+  }
+}
+
+/**
+ * Initialize the detector ID's and event type of the destination event lists.
+ * @param events : list of input events
+ * @param partials : resulting partial lists of events
+ */
+void TimeSplitter::initializePartials(const EventList &events, std::map<int, EventList *> partials) const {
+  // collect the state from events which is to be transferred to the partials
+
+  bool removeDetIDs{true};
+  auto detIDs = events.getDetectorIDs();
+  auto histogram = events.getHistogram();
+  auto eventType = events.getEventType();
+  // lambda expression initializing one partial
+  auto initPartial = [&](EventList *partial) {
+    partial->clear(removeDetIDs);
+    partial->setDetectorIDs(detIDs);
+    partial->setHistogram(histogram); // problem here!
+    partial->switchTo(eventType);
+  };
+  // iterate over the partials
+  for (auto iter = cbegin(partials); iter != cend(partials); ++iter)
```

Comment:
This is using constant iterators, and is exactly what [std::for_each](https://en.cppreference.com/w/cpp/algorithm/for_each) is meant to do. I think these lines should be
```c
std::for_each(partials.begin(), partials.end(), initPartial);
```